nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ListDatabaseTables.java

```diff
@@ -0,0 +1,304 @@
+        final ComponentLog logger = getLogger();
+        final DBCPService dbcpService = context.getProperty(DBCP_SERVICE).asControllerService(DBCPService.class);
+        final String catalog = context.getProperty(CATALOG).getValue();
+        final String schemaPattern = context.getProperty(SCHEMA_PATTERN).getValue();
+        final String tableNamePattern = context.getProperty(TABLE_NAME_PATTERN).getValue();
+        final String[] tableTypes = context.getProperty(TABLE_TYPES).isSet()
+                ? context.getProperty(TABLE_TYPES).getValue().split("\\s*,\\s*")
+                : null;
+        final boolean includeCount = context.getProperty(INCLUDE_COUNT).asBoolean();
+
+        final StateManager stateManager = context.getStateManager();
+        final StateMap stateMap;
+        final Map<String, String> stateMapProperties;
+        try {
+            stateMap = stateManager.getState(Scope.LOCAL);
+            stateMapProperties = new HashMap<>(stateMap.toMap());
+        } catch (IOException ioe) {
+            throw new ProcessException(ioe);
+        }
+
+        try (final Connection con = dbcpService.getConnection()) {
+
+            DatabaseMetaData dbMetaData = con.getMetaData();
+            ResultSet rs = dbMetaData.getTables(catalog, schemaPattern, tableNamePattern, tableTypes);
+            while (rs.next()) {
+                final String tableCatalog = rs.getString(1);
+                final String tableSchema = rs.getString(2);
+                final String tableName = rs.getString(3);
+                final String tableType = rs.getString(4);
+                final String tableRemarks = rs.getString(5);
+
+                // Build fully-qualified name
+                String fqn = Stream.of(tableCatalog, tableSchema, tableName)
+                        .filter(segment -> !StringUtils.isEmpty(segment))
+                        .collect(Collectors.joining("."));
+
+                String fqTableName = stateMap.get(fqn);
+                if (fqTableName == null) {
+                    FlowFile flowFile = session.create();
+                    logger.info("Found {}: {}", new Object[]{tableType, fqn});
+                    if (includeCount) {
+                        try (Statement st = con.createStatement()) {
+                            final String countQuery = "SELECT COUNT(1) FROM " + fqn;
+
+                            logger.debug("Executing query: {}", new Object[]{countQuery});
+                            ResultSet countResult = st.executeQuery(countQuery);
+                            if (countResult.next()) {
+                                flowFile = session.putAttribute(flowFile, DB_TABLE_COUNT, Long.toString(countResult.getLong(1)));
+                            }
+                        } catch (SQLException se) {
+                            logger.error("Couldn't get row count for {}", new Object[]{fqn});
+                            session.remove(flowFile);
```

Comment:
It's a data generator, so not sure what counts as data loss. I could add a "failure" relationship but I didn't see any precedence for transferring newly-created flowfiles.
If you mean that an error getting the count shouldn't prevent the flow file (minus the count) from being transferred, then yeah I can make that change. Just wasn't sure from a UX perspective if they want a count and can't get one, should the flow file be transferred?